code/modules/mob/living/simple_animal/hostile/necro.dm

```diff
@@ -96,11 +122,296 @@
+	Otherwise, start wandering and bust down some doors to find more food
+	*/
+	if(!isUnconscious())
+		if(stance == HOSTILE_STANCE_IDLE) //Not doing anything at the time
+			var/list/can_see = view(src, vision_range)
+			if(!busy)
+				if(can_evolve)//Can we evolve, and have we fed
+					busy = EVOLVING
+					check_evolve()
+				if((health < maxHealth) || (maxHealth < health_cap)) //Is there something to eat in range?
+					for(var/mob/living/carbon/human/C in can_see) //If so, chow down
+						if(C.stat == DEAD && !busy && check_edibility(C))
+							Goto(C, speed)
+							busy = MOVING_TO_TARGET
+							give_up(C) //If we're not there in 10 seconds, give up
+
+							if(C.Adjacent(src) && busy != EATING) //Once we've finally caught up
+								busy = EATING
+								eat(C)
+				if(!busy && break_doors)//So we don't try to eat and open doors
+					/*for(var/obj/machinery/light/L in can_see)
+						if(L.status != 1 && L.status !=2)
+							Goto(L, speed)
+							busy = MOVING_TO_TARGET
+							if(L.Adjacent(src) && busy != SMASHING_LIGHT)
+								L.attack_animal(src)*/
+
+					for(var/obj/machinery/door/D in can_see)//Else, let's open some doors and see what's around
+						if(can_open_door(D) && !busy)
+							Goto(D, speed)
+							busy = MOVING_TO_TARGET
+							give_up(D)
+
+							if(D.Adjacent(src) && busy != OPENING_DOOR)
+								busy = OPENING_DOOR
+								force_door(D)
+
+		else
+			busy = 0
+			stop_automated_movement = 0
+
+
+/mob/living/simple_animal/hostile/necro/zombie/proc/give_up(var/C)
+	spawn(100)
+		if(busy == MOVING_TO_TARGET)
+			if(target == C && get_dist(src,target) > 1)
+				target = null
+			busy = 0
+			stop_automated_movement = 0
+
+/mob/living/simple_animal/hostile/necro/zombie/proc/can_open_door(var/obj/machinery/door/D)
+	if(istype(D,/obj/machinery/door/poddoor) || istype(D, /obj/machinery/door/airlock/multi_tile/glass) || istype(D, /obj/machinery/door/window))
+		return 0
+
+	// Don't fuck with doors that are doing something
+	if(D.operating>0)
+		return 0
+
+	// Don't open opened doors.
+	if(!D.density)
+		return 0
+
+	// Can't open bolted/welded doors
+	if(istype(D,/obj/machinery/door/airlock))
+		var/obj/machinery/door/airlock/A=D
+		if(A.locked || A.welded || A.jammed)
+			if(break_doors == 2)
+				return 1
+			else
+				return 0
+
+	return 1
+
+/mob/living/simple_animal/hostile/necro/zombie/proc/force_door(var/obj/machinery/door/D)
+	stop_automated_movement = 1
+	D.visible_message("<span class='warning'>\The [D]'s motors whine as something attempts to brute force their way through it!</span>")
+	playsound(get_turf(D), 'sound/effects/grillehit.ogg', 50, 1)
+	D.shake(1, 8)
+	stop_automated_movement = 1
+	spawn(100) //Roughly 10 seconds for people on the other side of the door to run the heck away
+		if(can_open_door(D))//Let's see if nobody quickly bolted it
+			if(istype (src, /mob/living/simple_animal/hostile/necro/zombie/crimson)) //Guaranteed
```

Comment:
You should never be doing an istype check on src. Make a proc for the subtype to override.